pkg/gardenlet/controller/shoot/shoot/reconciler_delete.go

```diff
@@ -144,10 +148,10 @@ func (r *Reconciler) runDeleteShootFlow(ctx context.Context, o *operation.Operat
 
 	var (
 		nonTerminatingNamespace = botanist.SeedNamespaceObject.Status.Phase != corev1.NamespaceTerminating
-		cleanupShootResources   = nonTerminatingNamespace && kubeAPIServerDeploymentFound && infrastructure != nil
+		cleanupShootResources   = nonTerminatingNamespace && kubeAPIServerDeploymentFound && (infrastructure != nil || isWorkerless)
 		defaultInterval         = 5 * time.Second
 		defaultTimeout          = 30 * time.Second
-		staticNodesCIDR         = o.Shoot.GetInfo().Spec.Networking.Nodes != nil
+		staticNodesCIDR         = o.Shoot.GetInfo().Spec.Networking != nil && o.Shoot.GetInfo().Spec.Networking.Nodes != nil
```

Comment:
Side comment (not a review comment):
Go repeated a historic mistake of having nils which Rust avoided by removing the nil/null completely. So instead Spec.Networking 's type would be `Options<Networking>` whose values could be `None` (which means its not present) or `Some(n)`. This can then be chained safely to any level of nesting without the need for an explicit nil check.